.travis.yml

```diff
@@ -111,75 +134,77 @@ install:
     echo "repository hackage.haskell.org"               >> $CABALHOME/config
     echo "  url: http://hackage.haskell.org/"           >> $CABALHOME/config
   - |
-    if $GHCHEAD; then
+    if $HEADHACKAGE; then
     echo "allow-newer: $($HCPKG list --simple-output | sed -E 's/([a-zA-Z-]+)-[0-9.]+/*:\1/g')" >> $CABALHOME/config
-    
-    echo "repository head.hackage"                                                        >> $CABALHOME/config
-    echo "   url: http://head.hackage.haskell.org/"                                       >> $CABALHOME/config
+    echo "repository head.hackage.ghc.haskell.org"                                        >> $CABALHOME/config
+    echo "   url: https://ghc.gitlab.haskell.org/head.hackage/"                           >> $CABALHOME/config
     echo "   secure: True"                                                                >> $CABALHOME/config
-    echo "   root-keys: 07c59cb65787dedfaef5bd5f987ceb5f7e5ebf88b904bbd4c5cbdeb2ff71b740" >> $CABALHOME/config
-    echo "              2e8555dde16ebd8df076f1a8ef13b8f14c66bad8eafefd7d9e37d0ed711821fb" >> $CABALHOME/config
-    echo "              8f79fd2389ab2967354407ec852cbe73f2e8635793ac446d09461ffb99527f6e" >> $CABALHOME/config
+    echo "   root-keys: 7541f32a4ccca4f97aea3b22f5e593ba2c0267546016b992dfadcd2fe944e55d" >> $CABALHOME/config
+    echo "              26021a13b401500c8eb2761ca95c61f2d625bfef951b939a8124ed12ecf07329" >> $CABALHOME/config
+    echo "              f76d08be13e9a61a377a85e2fb63f4c5435d40f8feb3e12eb05905edb8cdea89" >> $CABALHOME/config
     echo "   key-threshold: 3"                                                            >> $CABALHOME/config
     fi
   - cat $CABALHOME/config
   - rm -fv cabal.project cabal.project.local cabal.project.freeze
   - travis_retry ${CABAL} v2-update -v
-  - if [ $HCNUMVER -eq 80801 ] ; then ${CABAL} v2-install -w ${HC} -j2 hlint --constraint='hlint ==2.1.*' | color_cabal_output ; fi
+  - if ! $GHCJS && [ $HCNUMVER -ge 80800 ] && [ $HCNUMVER -lt 80802 ] ; then ${CABAL} v2-install $WITHCOMPILER -j2 hlint --constraint='hlint ==2.1.*' | color_cabal_output ; fi
+  - if $GHCJS ; then ${CABAL} v2-install -w ghc-8.6.5 happy ; fi
```

Comment:
I'm curious to know why this line is necessary. (Disclaimer: I'm not at all familiar with GHCJS's best practices, so I apologize in advance if this is a dumb question.)

Reply:
We need to have happy (and actually all tools) to be available. cabal crosscompiling story is subpar. Currently it does build happy with GHCJS but cannot it use (and imo would be silly to).

Reply:
I'm still unclear on what exactly goes wrong if `happy` _isn't_ installed beforehand. Does a particular package fail to build?

Reply:
`haskell-src-exts` needs `happy`. And when we use GHCJS `cabal` fails to install&use proper `happy`. But things "just work" if there's `happy` in `.$PATH`.

I.e. there are bugs, but they luckily cancel each others out.

Reply:
I see! Has this bug been reported somewhere? I'm unclear if this is a bug with GHCJS, `cabal`, or both.

Reply:
Well... It's a mess. `Cabal` and `cabal-install` should treat GHCJS as cross-compiler, and even some ground work is done we need a man-year to get there.

`ghcjs-8.4` reports

```
 ,("Build platform","x86_64-unknown-linux")
 ,("Host platform","x86_64-unknown-linux")
 ,("Target platform","x86_64-unknown-linux")
```

which confuses `Cabal`. I think it's fixed in GHCJS-8.6 (or so I was told, cannot check, I don't have bindists), so maybe things works better with that.

That said, I don't dare to ask HVR to try to package 8.6 (only for this reason).

Reply:
OK. In that case, I'll keep this in mind when GHCJS-8.6 becomes widely available :)